ENH: Add SerializableImage class with to/from_bytes methods

effigies opened the pull request:
Closes #639.

Implements to/from_bytes for single-image variants of NIfTI-1, NIfTI-2, GIFTI and CIFTI2 images, as well as MGH images.

Created an abstract class deriving from `FileBasedImage` that can be used in other images, if it becomes desirable, but didn't want to spend a ton of time nailing it down for image types I have an, at best, passing familiarity with.

Shotgunosine commented:
to_bytes and from_bytes sounds good to me. Easier for a novice to interpret
as well.

On Tue, Jul 31, 2018 at 12:21 PM, Chris Markiewicz <notifications@github.com
> wrote:

> Current state: CIFTI issues. Not clear why it's not consistent with the
> file output, but may be due to to_file_map performing updates. Perhaps
> they're not idempotent.
>
> Based on TimVanMourik/Armadillo#11
> <https://github.com/TimVanMourik/Armadillo/issues/11>, it seems that
> deserializing would be similarly useful.
>
> Incidentally, I'm not committed to "(de)serialize" as function names. To
> be consistent with the rest of the API, we could use to_bytes and
> from_bytes. Let me know what y'all think.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/nipy/nibabel/pull/644#issuecomment-409281071>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AFC1mG82We8kHI6gSc4gTJf9nM9IyOdAks5uMIPvgaJpZM4U_0Of>
> .
>

effigies commented:
I think this is ready for a review, if anybody has some time. Pretty compact, in the end. I may not have used the testing machinery to its full potential, but hopefully it isn't too grotesque.

matthew-brett commented on nibabel/filebasedimages.py:
```diff
@@ -511,3 +512,79 @@ def path_maybe_image(klass, filename, sniff=None, sniff_max=1024):
+
+       * from_bytes(bytestring) - make instance by deserializing a byte string
+
+    The following properties should hold:
```
How do you want this to look in the rendered doc?  What happens with these indented lists?

effigies replied:
I actually haven't built the docs here. I'll fiddle with this.

matthew-brett commented on nibabel/filebasedimages.py:
```diff
@@ -511,3 +512,79 @@ def path_maybe_image(klass, filename, sniff=None, sniff_max=1024):
+
+    The following properties should hold:
+
+      * ``klass.from_bytes(bstr).to_bytes() == bstr``
```
Is that true in general for round-tripping?  For example, what happens with nifti images with image scaling?  I think the scale factors may get recalculated.

effigies replied:
Fair point. I guess what we want is load-level idempotence, to the same level of precision as save/load cycles.:

```Python
img_a = klass.from_bytes(bstr)
img_b = klass.from_bytes(img_a.to_bytes())

np.allclose(img_a.get_fdata(), img_b.get_fdata())
```

Would the following be too strong, as well?

```Python
img_a.to_bytes() == img_b.to_bytes()
```

matthew-brett commented on nibabel/filebasedimages.py:
```diff
@@ -511,3 +512,79 @@ def path_maybe_image(klass, filename, sniff=None, sniff_max=1024):
+    Further, for images that are single files on disk, the following methods of loading
+    the image must be equivalent:
+
+        img = klass.from_filename(fname)
```
Make this ``.. code:: python`` block?

matthew-brett commented on nibabel/filebasedimages.py:
```diff
@@ -511,3 +512,79 @@ def path_maybe_image(klass, filename, sniff=None, sniff_max=1024):
+        with open(fname, 'wb') as fobj:
+            fobj.write(img.to_bytes())
+
+    Images that consist of separate header and data files will generally
```
I'm not sure what this means - maybe give an example, of an Analyze image?

matthew-brett commented on nibabel/filebasedimages.py:
```diff
@@ -511,3 +512,79 @@ def path_maybe_image(klass, filename, sniff=None, sniff_max=1024):
+            Byte string containing the on-disk representation of an image
+        """
+        bio = io.BytesIO(bytestring)
+        file_map = klass.make_file_map({'image': bio, 'header': bio})
```
Ouch - this is a bit frightening.   I guess we now have to worry about the order in which the header and image get written, within the image classes?

effigies replied:
Yes. At present this works for single-file images, where they are already passed a single file object in both fields. I would expect that, if we extend to multi-file images, it would be the responsibility of the implementor to ensure that the serialization is consistent and can be de-serialized.

I'm a little hesitant to pre-engineer for those cases, as I don't have the time or need to work out those problems. As long as we're not providing this function for multi-image files, I'm not sure it's a problem.

effigies replied:
To be clear, if it looks like the decisions I've made now are going to actively make it harder to support those cases in the future than some other implementation, I'm happy to change it.

matthew-brett commented on nibabel/gifti/gifti.py:
```diff
@@ -724,6 +724,9 @@ def to_xml(self, enc='utf-8'):
 """ + xml.XmlSerializable.to_xml(self, enc)
 
+    # Avoid the indirection of going through to_file_map
+    to_bytes = to_xml
```
Just checking - `to_xml` does return a byte string?

effigies replied:
Yup. I checked.

matthew-brett commented on nibabel/tests/test_image_api.py:
```diff
@@ -451,6 +452,73 @@ def validate_affine_deprecated(self, imaker, params):
+        try:
+            header_eq = header_a == header_b
+        except NotImplementedError:
+            header_eq = type(header_a) == type(header_b)
```
Should we require a header equality check being defined?  This doesn't seem like a good test.

effigies replied:
This is a hack to compare `FileBasedHeader`s. I can put in a specific carve-out for that type, instead.

effigies commented:
@matthew-brett  I've addressed (or responded) to your comments, if you want to have another look.

effigies commented:
I've updated this PR to do the following:

1) The mixin methods `to_bytes`/`from_bytes` now check whether the class's `files_types` attribute indicates a multi-file image, and raises a `NotImplementedError`. This can be overridden by a multi-image class, but ensures we don't accidentally make promises we can't keep.
2) `Nifti1Pair` was marked serializable, when it should have been `Nifti1Image`. This has been corrected.
3) The header equality check in the API tests is now more specifically targeted to the one corner case, i.e., the abstract `FileBasedHeader` class.

@matthew-brett I believe this has addressed all of your comments. Please let me know if you have any further concerns.

I would like to get this in before the final Python 2 feature release (2.5.0) if possible.

If anybody else has some spare cycles, I'd appreciate a review on this one.

matthew-brett commented on nibabel/filebasedimages.py:
```diff
@@ -511,3 +512,89 @@ def path_maybe_image(klass, filename, sniff=None, sniff_max=1024):
+            fobj.write(img.to_bytes())
+
+    Images that consist of separate header and data files (e.g., Analyze
+    images) currently do not support this interface.
```
Maybe say something about the fact that image types can override this class method to implement it, for multi-file images.

matthew-brett commented on nibabel/filebasedimages.py:
```diff
@@ -511,3 +512,89 @@ def path_maybe_image(klass, filename, sniff=None, sniff_max=1024):
+    Images that consist of separate header and data files (e.g., Analyze
+    images) currently do not support this interface.
+    '''
+    @classmethod
```
PEP8 doesn't need a newline here?

matthew-brett commented on nibabel/tests/test_volumeutils.py:
```diff
@@ -700,7 +700,7 @@ def test_a2f_non_numeric():
     try:
         arr.astype(float)
-    except ValueError:
+    except (TypeError, ValueError):
```
Why do you need this change in this PR?

effigies replied:
Ah, nope. Rebased onto the wrong branch.

effigies commented:
Thanks for the review, Matthew.
